src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/HashAlgorithm.cs

```diff
@@ -99,24 +99,8 @@ public byte[] ComputeHash(Stream inputStream)
             if (_disposed)
                 throw new ObjectDisposedException(null);
 
-            // Use ArrayPool.Shared instead of CryptoPool because the array is passed out.
-            byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
-
-            int bytesRead;
-            int clearLimit = 0;
-
-            while ((bytesRead = inputStream.Read(buffer, 0, buffer.Length)) > 0)
-            {
-                if (bytesRead > clearLimit)
-                {
-                    clearLimit = bytesRead;
-                }
-
-                HashCore(buffer, 0, bytesRead);
-            }
+            inputStream.CopyTo((span, state) => ((HashAlgorithm)state!).HashCore(span), this, 4096);
```

Comment:
We should have made these overloads generic, the casting is such an eye sore.

Reply:
We generally avoid virtual generic methods.

Reply:
Yea, I'm aware but the downsides are:
- More casting, making the code uglier
- Unnecessary null checks (if you're passing a value type)
- The API looks less "modern"
- Boxing allocation when you want to pass a value type

Meh, it's unrelated to this change (which is a nice one)

Reply:
I agree with David.  Everything that uses (writable) SpanAction is a generic method over TState.

From a usability/design perspective, I'd make it generic.

If the whole point of this is to be more perfy and the generic method has a cost that exceeds the perfy wins, then I guess that'd be a reason to leave it as object (or just back out the feature).

Reply:
Are there any other places we will use this overload in dotnet/runtime?  Presumably not?  @davidfowl, are there any places ASP.NET will change to use this overload, given the current set of built-in streams that override it?  If we're going to close this PR and we're not going to use the method anywhere ourselves, my preference is to remove it entirely.

@jkotas, opinions on having this Stream virtual method be generic?

Reply:
Generic virtual methods have a cost profile similar to reflection. I do not think we want to use them on Stream.

I do not see a problem with casting. Casting is cheap operation and I believe that it is just fine for cases like these.

Reply:
> Are there any other places we will use this overload in dotnet/runtime? Presumably not? @davidfowl, are there any places ASP.NET will change to use this overload, given the current set of built-in streams that override it? If we're going to close this PR and we're not going to use the method anywhere ourselves, my preference is to remove it entirely.

Can't think of any off the top of my head but why remove it? There are lots of APIs we don't use ourselves but are generally useful (chicken and egg problem I guess).

> Generic virtual methods have a cost profile similar to reflection. I do not think we want to use them on Stream.

Can you quantify this? Especially since calls like `CopyTo` isn't as chatty as `Read` and I don't have a good sense for what you mean as far as "cost profile similar to reflection". 

> I do not see a problem with casting. Casting is cheap operation and I believe that it is just fine for cases like these.

I mean of course casting is fine its just ugly and using object means the API isn't type safe.

Reply:
> Can't think of any off the top of my head but why remove it?

Because if across our vast code bases we can't come up with a place where we'd use it, that suggests it's not particularly useful and we shouldn't pay the on-going costs of shipping it, of developers deciding whether or not to override it, and developers having to think through the trade-offs involved (e.g. a stream that doesn't override this, which will be pretty common, is going to have this method actually be more expensive than the existing overload).

Reply:
It's a factor but it can't be the only one. That said, this isn't one of those hills I want to die on. I was mostly commenting about the fact that whenever we add overloads that take callbacks with state, it should really be generic unless there are other factors (like being a virtual method) that outweigh the usability benefits.